### PR TITLE
App to callable function conversion rough draft

### LINKED ISSUE #571: WarningTreeWidget in qisis/apps/ipce and qisis/objs/ are duplicates
Also, the WarningTreeWidget in `qisis/objs` is being used in the Directory class but the on used for IPCE is not used at all.

### LINKED ISSUE #581: PhotoMetric Class defined twice in lrowacpho and lronacpho
The class should be moved into the objs folder.

### LINKED ISSUE #582: mocuncompress has funcions in xmain and pmain that are only used in main.cpp
So just move them to be local to main.cpp

### PR DESCRIPTION
**Pull request by Kelvinrr**
Changes:

* App directory source files containing `int main` are renamed to `main.cpp` so it's easier to distinguish the app code from library code. 
* CMake scripts updated to add apps as part of the ISIS3 library 
* closes #582 
* closes #581 
* Intermediate solution to #576, library code for those was moved into the objs folder for now 
* addresses #572, the two importfits were renamed to Nirs and Amica importfits since they are different enough and there is no obvious way to merge the two. 
* closes #571

### CODE REVIEW

File: isis/src/base/objs/Apps/Apps.truth [hunk 1]
```diff
@@ -0,0 +1 @@
+who cares
```
Comment (krlberry) [inline isis/src/base/objs/Apps/Apps.truth]: Nice truth data :+1:

File: isis/src/base/apps/crop/crop.cpp [hunk 2]
```diff
@@ -1,35 +1,37 @@
...
+  auto cropProccess = [&](Buffer &out)->void {
+    // Read the input line
+    int iline = sl + (out.Line() - 1) * linc;
+    in->SetLine(iline, sb);
+    cube->read(*in);
+
+    // Loop and move appropriate samples
+    for(int i = 0; i < out.size(); i++) {
+      out[i] = (*in)[(ss - 1) + i * sinc];
+    }
```
Comment (Kelvinrr) [inline isis/src/base/apps/crop/crop.cpp:27]: ProccessByLine/Brick can use lambdas with captures, this means the functions can be defined within the app func while capturing the required additional values as long as they are defined in the scope of the function. This avoids multiple packing/unpacking from using a UI object (which doesn't support arbitrary types) or void* array.

Comment (jessemapel) [reply isis/src/base/apps/crop/crop.cpp:27]: That sounds good to me

**Comment by Kelvinrr** [comment]
All mains of apps are located in a file named `main.cpp` located in any particular app folder.

File: isis/cmake/AddIsisModule.cmake [hunk 3]
```diff
@@ -99,10 +103,16 @@ function(add_isis_obj folder reqLibs)
 
   # Find the source and header files
   file(GLOB headers "${folder}/*.h" "${folder}/*.hpp")
+  # ignore app.cpp
   file(GLOB sources "${folder}/*.c" "${folder}/*.cpp")
   file(GLOB truths  "${folder}/*.truth")
   file(GLOB plugins "${folder}/*.plugin")
 
+  list(REMOVE_ITEM sources "${folder}/main.cpp")
```
Comment (Kelvinrr) [inline isis/cmake/AddIsisModule.cmake:108]: This makes sure main is not added to the library.

File: isis/src/base/apps/ckwriter/CkKernelWriter.cpp [hunk 4]
```diff
@@ -37,7 +37,7 @@
 #include "IString.h"
 #include "NaifStatus.h"
 #include "Pvl.h"
-#include "SpiceSegment.h"
+#include "CkSpiceSegment.h"
```
Comment (Kelvinrr) [inline isis/src/base/apps/ckwriter/CkKernelWriter.cpp:40]: The SpiceSegment class was separated into two different classes. Might quickly look into merging the two if possible.

File: isis/src/hayabusa/apps/amica2isis/AmicaImportFits.cpp [hunk 5]
```diff
@@ -19,7 +19,7 @@
  *   http://isis.astrogeology.usgs.gov, and the USGS privacy and disclaimers on
  *   http://www.usgs.gov/privacy.html.
  */
-#include "ImportFits.h"
+#include "AmicaImportFits.h"
```
Comment (Kelvinrr) [inline isis/src/hayabusa/apps/amica2isis/AmicaImportFits.cpp:22]: ImportFits was mostly duplicated between Amica and Nirs but there is a logicblock that is different. Need to consolidate the two and move into the ISIS3 lib.

File: isis/src/lro/apps/lrowacpho/Exponential.cpp [hunk 6]
```diff
@@ -60,12 +60,12 @@ namespace Isis {
      */
     double Exponential::photometry ( double i, double e, double g, int band ) const {
         // Test for valid band
-        if ((band <= 0) || (band > (int) _bandpho.size())) {
+        if ((band <= 0) || (band > (int) m_bandpho.size())) {
```
Comment (Kelvinrr) [inline isis/src/lro/apps/lrowacpho/Exponential.cpp:63]: Merged Photometery classes, only difference was naming conventions.

File: isis/src/mgs/apps/mocuncompress/main.cpp [hunk 7]
```diff
@@ -128,6 +148,303 @@ enum MocCompressEnum {
...
+/* Huffman tree (in table form) */
+uint8 code[256];
+uint8 left[256];
+uint8 right[256];
+
+int pred_past_eof;
+
+
+
+void decodeLoad(char *decodefile)
```
Comment (Kelvinrr) [inline isis/src/mgs/apps/mocuncompress/main.cpp:166]: Stuffed some floating functions in the app folder into main as they were only being used here.

File: isis/src/mro/apps/hi2isis/main.cpp [hunk 8]
```diff
@@ -36,6 +37,228 @@ int hisCount[6];
 int validCount[6];
 int section;
 
+
+// Construct a BLOb to contain the Hirise main line suffix and prefix data
+void SaveHiriseAncillaryData(ProcessImportPds &process, Cube *ocube) {
```
Comment (Kelvinrr) [inline isis/src/mro/apps/hi2isis/main.cpp:42]: Was originally floating in another .cpp file, only used in main so moved in to main.

**Comment by krlberry** [comment]
@Kelvinrr Could you update this with a new description? It doesn't need to go into a lot of detail -- I'm interested in a brief overview, if possible, and maybe which files contain the most important functional changes for this PR.

**Review by krlberry — CHANGES_REQUESTED** [review]
For some very straightforward pre-meeting feedback: 
`isis/src/mro/.DS_Store` and `isis/src/mro/apps/.DS_Store` should be removed from this PR.

File: isis/src/base/apps/crop/crop_func.cpp [hunk 9]
```diff
@@ -1,35 +1,37 @@
-#include "Isis.h"
```
Comment (Kelvinrr) [inline isis/src/base/apps/crop/crop_func.cpp:1]: rename to crop.cpp

File: isis/src/base/objs/ProcessByBrick/ProcessByBrick.cpp [hunk 10]
```diff
@@ -460,9 +461,50 @@ void ProcessByBrick::SetOutputRequirements(int outputRequirements) {
...
+   *
+   * @deprecated Please use ProcessCubeInPlace, ProcessCube, or ProcessCubes
+   * @param funct (Buffer &in) Receive an nxm brick in the input buffer.
+   *                                If n=1 and m=lines this will process by
+   *                                columns. Likewise if n=samples and m=1 this
+   *                                will process by lines.
+   *
+   * @throws iException::Programmer
+   */
+  void ProcessByBrick::StartProcess(std::function<void(Buffer &in)> funct ) {
```
Comment (Kelvinrr) [inline isis/src/base/objs/ProcessByBrick/ProcessByBrick.cpp:494]: Do this for in/out version too

File: isis/src/lro/apps/lrowacpho/HapkeExponential.cpp [hunk 11]
```diff
@@ -105,7 +105,7 @@ namespace Isis {
         //  surge term.
 
 //        I   µ0 + µ
-//        _ * ______  = A1 exp(B1 phase) + A2 ( µ0 + µ )exp(B2 phase) + A3 ( µ0 + µ ) + A4
+//        m_ * m______  = A1 exp(B1 phase) + A2 ( µ0 + µ )exp(B2 phase) + A3 ( µ0 + µ ) + A4
```
Comment (Kelvinrr) [inline isis/src/lro/apps/lrowacpho/HapkeExponential.cpp]: delete the m's

File: isis/src/mgs/apps/mocuncompress/getdecode.h [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+// #include "fs.h"
+// #include "predcode.h"
+//
+// typedef struct ht_node {
+//   int value;
+//   struct ht_node *zero, *one;
+// } Huffman_node;
+//
+// /* in TJL terminology, left is 0 and right is 1. */
+// extern uint8 code[], left[], right[];
```
Comment (Kelvinrr) [inline isis/src/mgs/apps/mocuncompress/getdecode.h]: Delete this file, make sure to re-test mocuncompress

**Comment by Kelvinrr** [comment]
Double check ZeroDark

**Review by krlberry — APPROVED** [review]

**Comment by jlaura** [comment]
@scsides Can you comment on this PR please.

**Comment by scsides** [comment]
@jlaura will add comments to issue #600

**Pull request merged**